### PR TITLE
Practitioner based report

### PR DESCRIPTION
**Pull request by maimoonak**
Fixes [#957 ](https://github.com/opensrp/fhir-resources/issues/957)

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/components/MeasureReportIndividualResultView.kt [hunk 1]
```diff
@@ -54,11 +54,11 @@ const val RESULT_VIEW_INDICATOR_DESCRIPTION = "resultViewIndicatorDescription"
 
 @Composable
 fun MeasureReportIndividualResultView(
-  modifier: Modifier = Modifier,
-  patientViewData: MeasureReportPatientViewData,
-  isMatchedIndicator: Boolean = true,
-  indicatorStatus: String = "",
-  indicatorDescription: String = ""
+    modifier: Modifier = Modifier,
+    patientViewData: MeasureReportSubjectViewData,
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/components/MeasureReportIndividualResultView.kt:58]: should we update the key too?
```suggestion
    subjectViewData: MeasureReportSubjectViewData,
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/components/MeasureReportIndividualResultView.kt [hunk 2]
```diff
@@ -132,7 +132,7 @@ fun MeasureReportIndividualResultView(
 private fun MeasureReportIndividualResultViewPreview() {
   MeasureReportIndividualResultView(
     patientViewData =
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/components/MeasureReportIndividualResultView.kt:134]:
```suggestion
    subjectViewData =
```

Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/components/MeasureReportIndividualResultView.kt:134]:
```suggestion
    subjectViewData =
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt [hunk 3]
```diff
@@ -75,14 +75,14 @@ fun MeasureReportResultScreen(
...
-  endDate: String,
-  patientViewData: MeasureReportPatientViewData?,
-  measureReportIndividualResult: MeasureReportIndividualResult?,
-  measureReportPopulationResult: List<MeasureReportPopulationResult>?,
-  modifier: Modifier = Modifier
+    screenTitle: String,
+    navController: NavController,
+    startDate: String,
+    endDate: String,
+    patientViewData: MeasureReportSubjectViewData?,
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt:82]:
```suggestion
    subjectViewData: MeasureReportSubjectViewData?,
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt [hunk 4]
```diff
@@ -147,7 +147,7 @@ private fun MeasureReportResultScreenForIndividualPreview() {
     startDate = "25 Nov, 2021",
     endDate = "29 Nov, 2021",
     patientViewData =
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt:149]:
```suggestion
    subjectViewData =
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt [hunk 5]
```diff
@@ -172,7 +172,7 @@ private fun MeasureReportResultScreenForPopulationPreview() {
...
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/screens/MeasureReportResultScreen.kt:174]:
```suggestion
    subjectViewData =
```

**Review by pld — COMMENTED** [review]
I think I get what you're planning here, can you add an issue please or link this to the issue it's addressing, thank you!

**Review by pld — DISMISSED** [review]
it looks like the tests need to be adapted

```

> Task :quest:compileQuestDebugKotlin FAILED
e: /Users/runner/work/fhircore/fhircore/android/quest/src/main/java/org/smartregister/fhircore/quest/data/report/measure/MeasureReportRepository.kt: (55, 11): Cannot find a parameter with this name: baseResource
e: /Users/runner/work/fhircore/fhircore/android/quest/src/main/java/org/smartregister/fhircore/quest/data/report/measure/MeasureReportRepository.kt: (56, 11): Cannot find a parameter with this name: relatedRepositoryResourceData
e: /Users/runner/work/fhircore/fhircore/android/quest/src/main/java/org/smartregister/fhircore/quest/data/report/measure/MeasureReportRepository.kt: (56, 43): Not enough information to infer type variable E
e: /Users/runner/work/fhircore/fhircore/android/quest/src/main/java/org/smartregister/fhircore/quest/data/report/measure/MeasureReportRepository.kt: (59, 9): No value passed for parameter 'repositoryResourceData'
```

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt [hunk 6]
```diff
@@ -335,28 +350,36 @@ constructor(
...
-            // MeasureEvaluator looks for Group members and skips the Group itself
-            if (it is Group && !it.hasMember()) {
-              it.addMember(Group.GroupMemberComponent(it.asReference()))
-              fhirEngine.update(it)
-            }
-            runMeasureReport(measureUrl, SUBJECT, startDateFormatted, endDateFormatted, subject)
-          }
+      if (subjects.isNotEmpty()) {
+        subjects
+          .map { runMeasureReport(measureUrl, SUBJECT, startDateFormatted, endDateFormatted, it) }
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt:382]: this looks good, at a later date we can experiment with using mapp here, but totally possible that running in parallel might actually lead to worse performance

**Review by pld — COMMENTED** [review]
this looks great! lots of great code to get multi-subject reporting working!

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt [hunk 7]
```diff
@@ -335,28 +350,36 @@ constructor(
...
 
+  private suspend fun fetchSubjects(config: MeasureReportConfig): List<String> {
+    return reportTypeSelectorUiState.value.subjectViewData
+      .ifEmpty {
+        if (config.subjectXFhirQuery?.isNotEmpty() == true) {
+          fhirEngine.search(config.subjectXFhirQuery!!).map {
+            // TODO a hack to prevent missing subject in case of Group based reports where
+            // MeasureEvaluator looks for Group members and skips the Group itself
+            if (it is Group && !it.hasMember()) {
+              it.addMember(Group.GroupMemberComponent(it.asReference()))
```
[the remark this answers is not in the document]
Comment (maimoonak) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt:361]: Yeah. The SubjectEvalutor just works on Group.member. And Group itself is ignored. org.opencds.cqf.cql.evaluator.measure.r4.R4FhirDalSubjectProvider L52.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt [hunk 8]
```diff
@@ -85,6 +87,14 @@ fun Base?.valueToString(): String {
       this.given.firstOrNull().let {
         (if (it != null) "${it.valueToString()} " else "").plus(this.family)
       }
+    this is Patient ->
+      this.nameFirstRep.nameAsSingleString +
+        ", " +
+        this.gender.name.first() +
+        ", " +
+        this.birthDate.yearsPassed()
+    this is Practitioner -> this.nameFirstRep.nameAsSingleString + ", " + this.gender?.name?.first()
```
[the remark this answers is not in the document]
Comment (pld) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt:96]: k cool, I think we should remove for now, maybe in future we make this configurable or revisit

```suggestion
    this is Practitioner -> this.nameFirstRep.nameAsSingleString
```

**Review by pld — DISMISSED** [review]

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt (hunk 6 above)
Comment (ndegwamartin) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt:382]: Yeah I think we can try this and see

**Review by ndegwamartin — APPROVED** [review]

**Pull request merged**